### ISSUE TITLE
Implement parsimony pressure

### ISSUE THREAD

**Issue by piprrr** [AureumChaos/LEAP#27]
Variable length genomes typically suffer from bloat, and the common solution is to impose parsimony pressure by penalizing fitness based on genome length.

This also provides an excellent use case for resolving a critical design issue of determining responsibility for comparing two or more individuals. The current design does not easily accommodate parsimony pressure because the only signals used for comparison is fitness, and there is no easy way to have genome length contribute to the final comparison.

**Comment by piprrr** [comment]
This will entail rejiggering the comparison as a standalone.

**Comment by SigmaX** [comment]
Discussion from today's meeting:

- Do this by adding a `key` option to **selection operators**, to permit customizing the default comparison based on `<`/`>`.
  - Solved our use cases, because operators receive `Individual` objects.  So the `key` function can access genomes or whatever else it needs.
  - This has the added virtue of making it easy to, say, plot the vanilla fitness, but do selection on the a penalized fitness.
  - We might want to also add the `key` option to probes, so we have the option of plotting penalized fitness too.
  
- Also a good time to fix **floating point comparisons** in our fitness `__eq__()` logic.

We've also discussed possible refactorings, like splitting out a `FitnessComparator` separately from `Problem`.  Tabled for now (but potentially relevant for multi-objective?).

**Comment by piprrr** [comment]
Turns out our clever plan wasn't so clever.  The optional `key` argument seen in such functions as `list.sort()` take a *single argument*, and one of that type already being sorted.  E.g., a list of strings could have a `key` of `str.lowercase` passed in that would lowercase all the strings first and then do the sort.  Translated to our domain, selection is sorting on fitnesses, so `key` arguments *would only work on fitnesses*.  Therefore this approach is a non-starter because we also need to look at genome lengths in addition to fitnesses for parsimony pressure.

**Comment by SigmaX** [comment]
That doesn't sound right to me.  The classic pattern for `key` is to pass it `itemgetter(1), ex. to sort on a certain element of a tuple---and that involves sorting on a different type than the intput:

```python
from operator import itemgetter

l = [('a', 15), ('b', 12), ('c', 1), ('e', 21), ('d', 21)]
sorted(l, key=itemgetter(1))
# Result:
#  [('c', 1), ('b', 12), ('a', 15), ('e', 21), ('d', 21)]
```

I tried out a lexicographic example of truncation selection in a Jupyter notebook.  This seems to work fine:

```python
import toolz

from leap_ec.individual import Individual
from leap_ec.real_rep.problems import SpheroidProblem

problem = SpheroidProblem(maximize=True)


pop = [ Individual([ 1, 2, 3, 4 ], problem=problem),
        Individual([ 1, 2, 3 ], problem=problem),
        Individual([ 1, 4 ], problem=problem),
        Individual([ 1, 2, 3, 4, 5 ], problem=problem) ]
pop[0].fitness = 0.4
pop[1].fitness = 1.0
pop[2].fitness = 0.4
pop[3].fitness = 0.4

pop = iter(pop)

toolz.topk(2, pop, key=lambda x: (x.fitness, -len(x.genome)))

# Result:
# (Individual([1, 2, 3], None, <leap_ec.real_rep.problems.SpheroidProblem object at 0x7f71fb61a220>),
#  Individual([1, 4], None, <leap_ec.real_rep.problems.SpheroidProblem object at 0x7f71fb61a220>))
```

**Comment by SigmaX** [comment]
I think I misunderstood your point.

But regardless, our selection operators sort `Individual`s, not fitnesses, so we can pass in things like `lambda x: (x.fitness, -len(x.genome))`.

**Comment by piprrr** [comment]
Yes, we can, and that will work with regular ole parsimony pressure where we ding each individual based on the genome.  This breaks down for lexicographical parsimony pressure because you need to see *both* individuals to do the tie breaker based on length.  :(

It may be that we have *two different* parsimony solutions.  One that subclasses, as happens with Legacy LEAP, to do the lexicographic parsimony, and another that allows for passing in a key to override the default fitness based one.  I think that gives us the most flexibility.

**Comment by SigmaX** [comment]
Actually the example I gave above does lexicographic: comparison of Tuples in Python is lexicographic, so returning a tuple from the lambda gives us what we need.

I agree that being able to pass in a comparator would offer more flexibility.  Not every kind of comparison can be reduced to a total ordering on the fitness values (which is what the key mechanism forces us into).

But I think a total ordering covers all the use cases we have discussed?
